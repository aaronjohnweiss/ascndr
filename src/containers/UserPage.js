import React, {Fragment} from 'react'
import {connect} from 'react-redux'
import Button from 'react-bootstrap/Button'
import EntityModal from '../components/EntityModal'
import {Col, ListGroup, Row} from 'react-bootstrap'
import {addFriendFields, userFields, userIdValidation, userNameValidation} from '../templates/userFields'
import {compose} from 'redux'
import {firebaseConnect, isLoaded} from 'react-redux-firebase'
import {distinct, findUser, findUserKey, getFriendsForUser, getUserName} from '../helpers/filterUtils';
import ConfirmCancelButton from "../components/ConfirmCancelButton";
import {useModalState} from "../helpers/useModalState";
import {ActivityCalendarSettingsModal} from "../components/ActivityCalendarSettingsModal";

const UserPage = ({auth: {uid}, users, firebase}) => {
    const [showUserModal, openUserModal, closeUserModal] = useModalState(false);

    const [showFriendModal, openFriendModal, closeFriendModal] = useModalState(false);

    const [showCalendarModal, openCalendarModal, closeCalendarModal] = useModalState(false);

    if (!isLoaded(users)) return 'Loading'

    const userInfo = findUser(users, uid);

    const userName = userInfo && userInfo.name;

    const friends = getFriendsForUser(userInfo, users);

    const handleUserUpdate = user => {
        const key = findUserKey(users, uid);

        const data = {...userInfo, ...user, uid};
        if (key) {
            firebase.update(`users/${key}`, data);
        } else {
            firebase.push('users', data);
        }
        closeUserModal();
    }

    const addFriend = ({userValue}) => {
        let uidToAdd
        if (users.map(user => user.value.uid).includes(userValue)) {
            uidToAdd = userValue
        } else {
            uidToAdd = (users.map(user => user.value).find(user => user.name === userValue) || {}).uid
        }


        if (uidToAdd !== uid) {
            handleUserUpdate({...userInfo, friends: distinct([...(userInfo.friends || []), uidToAdd])})
        }
        closeFriendModal();
    }

    const removeFriend = uidToRemove => {
        handleUserUpdate({...userInfo, friends: (userInfo.friends || []).filter(friendUid => friendUid !== uidToRemove)});
        closeFriendModal();
    }

    const setPreferences = name => preferences => {
        handleUserUpdate({...userInfo, preferences: {...userInfo.preferences, [name]: preferences}, uid})
        closeCalendarModal()
    }

    return (
        <Fragment>
            <p>Your uid: <b>{uid}</b></p>
            <div className="align-items-center">
                <Row className="align-items-center">
                    <Col>
                        <p className="align-middle m-0">Your display name: {userName ? <b>{userName}</b> :
                            <i>Not set</i>}</p>
                    </Col>
                    <Col xs="auto">
                        <Button className="float-end" onClick={openUserModal}>Edit User</Button>
                    </Col>
                </Row>

            </div>
            <h3>Preferences</h3>
            <ListGroup>
                <ListGroup.Item className='align-items-center'>
                    <div className='align-middle float-start'>Activity Calendar Settings</div>
                    <Button size='sm' onClick={openCalendarModal} className='float-end'>Edit</Button>
                </ListGroup.Item>
            </ListGroup>

            <h3>Friends</h3>
            <ListGroup>
                {friends.map((friend, idx) => (
                    <ListGroup.Item key={idx} className='align-items-center'>
                        <div className='align-middle float-start'>{getUserName(friend)}</div>
                        <ConfirmCancelButton buttonText='-' buttonProps={{size: 'sm', className: 'float-end'}}
                                             handleConfirm={() => removeFriend(friend.uid)} modalTitle='Remove friend'
                                             modalBody={`Remove ${getUserName(friend)} from your friends list?`}/>
                    </ListGroup.Item>
                ))}
            </ListGroup>
            <br/>
            <div className="d-grid d-block">
                <Button variant='primary' onClick={openFriendModal}>
                    Add Friend
                </Button>
            </div>

            <EntityModal show={showUserModal}
                         title='Edit User'
                         handleClose={closeUserModal}
                         handleSubmit={handleUserUpdate}
                         fields={userFields}
                         validateState={userNameValidation(users, userName)}
                         initialValues={{...userInfo}}/>

            <EntityModal show={showFriendModal}
                         title='Add friend'
                         handleClose={closeFriendModal}
                         handleSubmit={addFriend}
                         fields={addFriendFields}
                         validateState={userIdValidation(users)}/>

            <ActivityCalendarSettingsModal show={showCalendarModal} user={userInfo} friends={friends}
                                           handleSubmit={setPreferences('activityCalendar')} handleClose={closeCalendarModal}/>
        </Fragment>
    )
}

const mapStateToProps = state => {
    return {
        users: state.firebase.ordered.users,
        auth: state.auth
    }
}

export default compose(
    firebaseConnect([
        {path: 'users'}
    ]),
    connect(mapStateToProps)
)(UserPage)