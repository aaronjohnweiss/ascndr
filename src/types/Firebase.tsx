export type Persisted<T> = {key: string, value: T}
export type OrderedList<T> = Persisted<T>[]
export type Data<T> = Record<string, T>