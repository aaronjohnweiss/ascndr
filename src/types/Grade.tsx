import { ALL_MODIFIERS, ALL_STYLES } from '../helpers/gradeUtils'

export type RouteStyle = (typeof ALL_STYLES)[number]
export type RouteModifier = (typeof ALL_MODIFIERS)[number]

export const isStyle = (str: string): str is RouteStyle =>
  ALL_STYLES.find(style => style === str) !== undefined
export interface Grade {
  difficulty: number
  style: RouteStyle
  modifier?: RouteModifier
}

export interface GradeDecoration {
  percentage?: number
}

export type DecoratedGrade = Grade & GradeDecoration

export type DecoratedCustomGrade = GradeDecoration & { key: string }
