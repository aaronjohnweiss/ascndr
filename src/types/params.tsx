export type IdParams = {
  id: string
}
