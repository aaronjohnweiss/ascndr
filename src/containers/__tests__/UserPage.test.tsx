import {
  defaultState,
  mockFirebase,
  mountWithProviders,
  setupMockFirebase,
  user,
  userKey,
} from '../../redux/__tests__/reducer.test'
import UserPage from '../UserPage'
import { defaultUser } from '../../types/User'

describe('UserPage', () => {
  beforeEach(() => {
    setupMockFirebase()
  })

  it('should render', () => {
    expect(mountWithProviders(UserPage).component).toMatchSnapshot()
  })

  it('should allow adding a friend', () => {
    const state = { ...defaultState }
    const newFriendUid = 'newFriend'
    state.firebase.ordered.users.push({
      key: 'newFriendKey',
      value: defaultUser({ uid: newFriendUid }),
    })

    const wrapper = mountWithProviders(UserPage, { reduxParams: { preloadedState: state } }).wrapper

    wrapper.find('Button#add-friend').simulate('click')
    const modal = wrapper.find('EntityModal[title="Add friend"]')
    modal.find('input#userValue').simulate('change', { target: { value: newFriendUid } })
    modal.find('Button#modal-submit').simulate('click')

    expect(mockFirebase.update).toHaveBeenCalledWith(
      `users/${userKey}`,
      expect.objectContaining({
        ...user,
        friends: [...user.friends, newFriendUid],
      }),
    )
  })

  it('should allow updating user', () => {
    const newName = 'New user name'

    const wrapper = mountWithProviders(UserPage).wrapper

    wrapper.find('Button#edit-user').simulate('click')
    const modal = wrapper.find('EntityModal[title="Edit User"]')
    modal.find('input#name').simulate('change', { target: { value: newName } })
    modal.find('Button#modal-submit').simulate('click')

    expect(mockFirebase.update).toHaveBeenCalledWith(
      `users/${userKey}`,
      expect.objectContaining({
        ...user,
        name: newName,
      }),
    )
  })
})
