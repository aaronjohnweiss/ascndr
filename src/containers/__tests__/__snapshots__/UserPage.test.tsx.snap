// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage should render 1`] = `
<UserPage
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
>
  <p>
    Your uid: 
    <b>
      myUid
    </b>
  </p>
  <div
    className="align-items-center"
  >
    <Row
      className="align-items-center"
    >
      <div
        className="align-items-center row"
      >
        <Col>
          <div
            className="col"
          >
            <p
              className="align-middle m-0"
            >
              Your display name: 
              <b>
                Mr. User
              </b>
            </p>
          </div>
        </Col>
        <Col
          xs="auto"
        >
          <div
            className="col-auto"
          >
            <Button
              active={false}
              className="float-end"
              disabled={false}
              id="edit-user"
              onClick={[Function]}
              variant="primary"
            >
              <button
                className="float-end btn btn-primary"
                disabled={false}
                id="edit-user"
                onClick={[Function]}
                type="button"
              >
                Edit User
              </button>
            </Button>
          </div>
        </Col>
      </div>
    </Row>
  </div>
  <h3>
    Preferences
  </h3>
  <ListGroup>
    <ForwardRef
      as="div"
      className="list-group"
      onSelect={[Function]}
    >
      <div
        className="list-group"
        onKeyDown={[Function]}
      >
        <ListGroupItem
          active={false}
          className="align-items-center"
          disabled={false}
        >
          <ForwardRef
            as="div"
            className="align-items-center list-group-item"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="align-items-center list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="align-middle float-start"
              >
                Activity Calendar Settings
              </div>
              <Button
                active={false}
                className="float-end"
                disabled={false}
                onClick={[Function]}
                size="sm"
                variant="primary"
              >
                <button
                  className="float-end btn btn-primary btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Edit
                </button>
              </Button>
            </div>
          </ForwardRef>
        </ListGroupItem>
      </div>
    </ForwardRef>
  </ListGroup>
  <h3>
    Friends
  </h3>
  <ListGroup>
    <ForwardRef
      as="div"
      className="list-group"
      onSelect={[Function]}
    >
      <div
        className="list-group"
        onKeyDown={[Function]}
      >
        <ListGroupItem
          active={false}
          className="align-items-center"
          disabled={false}
          id="uid0"
          key="0"
        >
          <ForwardRef
            as="div"
            className="align-items-center list-group-item"
            disabled={false}
            id="uid0"
            onClick={[Function]}
          >
            <div
              className="align-items-center list-group-item"
              data-rb-event-key={null}
              disabled={false}
              id="uid0"
              onClick={[Function]}
            >
              <div
                className="align-middle float-start"
              >
                user 0
              </div>
              <ConfirmCancelButton
                buttonProps={
                  Object {
                    "className": "float-end",
                    "size": "sm",
                  }
                }
                buttonText="-"
                handleConfirm={[Function]}
                modalBody="Remove user 0 from your friends list?"
                modalTitle="Remove friend"
              >
                <WrappedButton>
                  <ShowModalButton>
                    <Button
                      active={false}
                      className="float-end"
                      disabled={false}
                      onClick={[Function]}
                      size="sm"
                      variant="danger"
                    >
                      <button
                        className="float-end btn btn-danger btn-sm"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        -
                      </button>
                    </Button>
                  </ShowModalButton>
                </WrappedButton>
                <ConfirmCancelModal
                  handleConfirm={[Function]}
                  hideModal={[Function]}
                  modalBody="Remove user 0 from your friends list?"
                  modalTitle="Remove friend"
                  showModal={false}
                >
                  <Modal
                    animation={true}
                    autoFocus={true}
                    backdrop={true}
                    dialogAs={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "ModalDialog",
                        "render": [Function],
                      }
                    }
                    enforceFocus={true}
                    keyboard={true}
                    onHide={[Function]}
                    restoreFocus={true}
                    show={false}
                  >
                    <Modal
                      autoFocus={true}
                      backdrop={true}
                      backdropTransition={[Function]}
                      containerClassName="modal-open"
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        BootstrapModalManager {
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "modals": Array [],
                          "scrollbarSize": 0,
                        }
                      }
                      onEnter={[Function]}
                      onEntering={[Function]}
                      onEscapeKeyDown={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      renderDialog={[Function]}
                      restoreFocus={true}
                      show={false}
                      transition={[Function]}
                    />
                  </Modal>
                </ConfirmCancelModal>
              </ConfirmCancelButton>
            </div>
          </ForwardRef>
        </ListGroupItem>
        <ListGroupItem
          active={false}
          className="align-items-center"
          disabled={false}
          id="uid1"
          key="1"
        >
          <ForwardRef
            as="div"
            className="align-items-center list-group-item"
            disabled={false}
            id="uid1"
            onClick={[Function]}
          >
            <div
              className="align-items-center list-group-item"
              data-rb-event-key={null}
              disabled={false}
              id="uid1"
              onClick={[Function]}
            >
              <div
                className="align-middle float-start"
              >
                user 1
              </div>
              <ConfirmCancelButton
                buttonProps={
                  Object {
                    "className": "float-end",
                    "size": "sm",
                  }
                }
                buttonText="-"
                handleConfirm={[Function]}
                modalBody="Remove user 1 from your friends list?"
                modalTitle="Remove friend"
              >
                <WrappedButton>
                  <ShowModalButton>
                    <Button
                      active={false}
                      className="float-end"
                      disabled={false}
                      onClick={[Function]}
                      size="sm"
                      variant="danger"
                    >
                      <button
                        className="float-end btn btn-danger btn-sm"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        -
                      </button>
                    </Button>
                  </ShowModalButton>
                </WrappedButton>
                <ConfirmCancelModal
                  handleConfirm={[Function]}
                  hideModal={[Function]}
                  modalBody="Remove user 1 from your friends list?"
                  modalTitle="Remove friend"
                  showModal={false}
                >
                  <Modal
                    animation={true}
                    autoFocus={true}
                    backdrop={true}
                    dialogAs={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "ModalDialog",
                        "render": [Function],
                      }
                    }
                    enforceFocus={true}
                    keyboard={true}
                    onHide={[Function]}
                    restoreFocus={true}
                    show={false}
                  >
                    <Modal
                      autoFocus={true}
                      backdrop={true}
                      backdropTransition={[Function]}
                      containerClassName="modal-open"
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        BootstrapModalManager {
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "modals": Array [],
                          "scrollbarSize": 0,
                        }
                      }
                      onEnter={[Function]}
                      onEntering={[Function]}
                      onEscapeKeyDown={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      renderDialog={[Function]}
                      restoreFocus={true}
                      show={false}
                      transition={[Function]}
                    />
                  </Modal>
                </ConfirmCancelModal>
              </ConfirmCancelButton>
            </div>
          </ForwardRef>
        </ListGroupItem>
        <ListGroupItem
          active={false}
          className="align-items-center"
          disabled={false}
          id="uid2"
          key="2"
        >
          <ForwardRef
            as="div"
            className="align-items-center list-group-item"
            disabled={false}
            id="uid2"
            onClick={[Function]}
          >
            <div
              className="align-items-center list-group-item"
              data-rb-event-key={null}
              disabled={false}
              id="uid2"
              onClick={[Function]}
            >
              <div
                className="align-middle float-start"
              >
                user 2
              </div>
              <ConfirmCancelButton
                buttonProps={
                  Object {
                    "className": "float-end",
                    "size": "sm",
                  }
                }
                buttonText="-"
                handleConfirm={[Function]}
                modalBody="Remove user 2 from your friends list?"
                modalTitle="Remove friend"
              >
                <WrappedButton>
                  <ShowModalButton>
                    <Button
                      active={false}
                      className="float-end"
                      disabled={false}
                      onClick={[Function]}
                      size="sm"
                      variant="danger"
                    >
                      <button
                        className="float-end btn btn-danger btn-sm"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        -
                      </button>
                    </Button>
                  </ShowModalButton>
                </WrappedButton>
                <ConfirmCancelModal
                  handleConfirm={[Function]}
                  hideModal={[Function]}
                  modalBody="Remove user 2 from your friends list?"
                  modalTitle="Remove friend"
                  showModal={false}
                >
                  <Modal
                    animation={true}
                    autoFocus={true}
                    backdrop={true}
                    dialogAs={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "ModalDialog",
                        "render": [Function],
                      }
                    }
                    enforceFocus={true}
                    keyboard={true}
                    onHide={[Function]}
                    restoreFocus={true}
                    show={false}
                  >
                    <Modal
                      autoFocus={true}
                      backdrop={true}
                      backdropTransition={[Function]}
                      containerClassName="modal-open"
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        BootstrapModalManager {
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "modals": Array [],
                          "scrollbarSize": 0,
                        }
                      }
                      onEnter={[Function]}
                      onEntering={[Function]}
                      onEscapeKeyDown={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      renderDialog={[Function]}
                      restoreFocus={true}
                      show={false}
                      transition={[Function]}
                    />
                  </Modal>
                </ConfirmCancelModal>
              </ConfirmCancelButton>
            </div>
          </ForwardRef>
        </ListGroupItem>
        <ListGroupItem
          active={false}
          className="align-items-center"
          disabled={false}
          id="uid3"
          key="3"
        >
          <ForwardRef
            as="div"
            className="align-items-center list-group-item"
            disabled={false}
            id="uid3"
            onClick={[Function]}
          >
            <div
              className="align-items-center list-group-item"
              data-rb-event-key={null}
              disabled={false}
              id="uid3"
              onClick={[Function]}
            >
              <div
                className="align-middle float-start"
              >
                user 3
              </div>
              <ConfirmCancelButton
                buttonProps={
                  Object {
                    "className": "float-end",
                    "size": "sm",
                  }
                }
                buttonText="-"
                handleConfirm={[Function]}
                modalBody="Remove user 3 from your friends list?"
                modalTitle="Remove friend"
              >
                <WrappedButton>
                  <ShowModalButton>
                    <Button
                      active={false}
                      className="float-end"
                      disabled={false}
                      onClick={[Function]}
                      size="sm"
                      variant="danger"
                    >
                      <button
                        className="float-end btn btn-danger btn-sm"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        -
                      </button>
                    </Button>
                  </ShowModalButton>
                </WrappedButton>
                <ConfirmCancelModal
                  handleConfirm={[Function]}
                  hideModal={[Function]}
                  modalBody="Remove user 3 from your friends list?"
                  modalTitle="Remove friend"
                  showModal={false}
                >
                  <Modal
                    animation={true}
                    autoFocus={true}
                    backdrop={true}
                    dialogAs={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "ModalDialog",
                        "render": [Function],
                      }
                    }
                    enforceFocus={true}
                    keyboard={true}
                    onHide={[Function]}
                    restoreFocus={true}
                    show={false}
                  >
                    <Modal
                      autoFocus={true}
                      backdrop={true}
                      backdropTransition={[Function]}
                      containerClassName="modal-open"
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        BootstrapModalManager {
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "modals": Array [],
                          "scrollbarSize": 0,
                        }
                      }
                      onEnter={[Function]}
                      onEntering={[Function]}
                      onEscapeKeyDown={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      renderDialog={[Function]}
                      restoreFocus={true}
                      show={false}
                      transition={[Function]}
                    />
                  </Modal>
                </ConfirmCancelModal>
              </ConfirmCancelButton>
            </div>
          </ForwardRef>
        </ListGroupItem>
        <ListGroupItem
          active={false}
          className="align-items-center"
          disabled={false}
          id="uid4"
          key="4"
        >
          <ForwardRef
            as="div"
            className="align-items-center list-group-item"
            disabled={false}
            id="uid4"
            onClick={[Function]}
          >
            <div
              className="align-items-center list-group-item"
              data-rb-event-key={null}
              disabled={false}
              id="uid4"
              onClick={[Function]}
            >
              <div
                className="align-middle float-start"
              >
                user 4
              </div>
              <ConfirmCancelButton
                buttonProps={
                  Object {
                    "className": "float-end",
                    "size": "sm",
                  }
                }
                buttonText="-"
                handleConfirm={[Function]}
                modalBody="Remove user 4 from your friends list?"
                modalTitle="Remove friend"
              >
                <WrappedButton>
                  <ShowModalButton>
                    <Button
                      active={false}
                      className="float-end"
                      disabled={false}
                      onClick={[Function]}
                      size="sm"
                      variant="danger"
                    >
                      <button
                        className="float-end btn btn-danger btn-sm"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        -
                      </button>
                    </Button>
                  </ShowModalButton>
                </WrappedButton>
                <ConfirmCancelModal
                  handleConfirm={[Function]}
                  hideModal={[Function]}
                  modalBody="Remove user 4 from your friends list?"
                  modalTitle="Remove friend"
                  showModal={false}
                >
                  <Modal
                    animation={true}
                    autoFocus={true}
                    backdrop={true}
                    dialogAs={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "ModalDialog",
                        "render": [Function],
                      }
                    }
                    enforceFocus={true}
                    keyboard={true}
                    onHide={[Function]}
                    restoreFocus={true}
                    show={false}
                  >
                    <Modal
                      autoFocus={true}
                      backdrop={true}
                      backdropTransition={[Function]}
                      containerClassName="modal-open"
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        BootstrapModalManager {
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "modals": Array [],
                          "scrollbarSize": 0,
                        }
                      }
                      onEnter={[Function]}
                      onEntering={[Function]}
                      onEscapeKeyDown={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      renderDialog={[Function]}
                      restoreFocus={true}
                      show={false}
                      transition={[Function]}
                    />
                  </Modal>
                </ConfirmCancelModal>
              </ConfirmCancelButton>
            </div>
          </ForwardRef>
        </ListGroupItem>
      </div>
    </ForwardRef>
  </ListGroup>
  <br />
  <div
    className="d-grid d-block"
  >
    <Button
      active={false}
      disabled={false}
      id="add-friend"
      onClick={[Function]}
      variant="primary"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        id="add-friend"
        onClick={[Function]}
        type="button"
      >
        Add Friend
      </button>
    </Button>
  </div>
  <EntityModal
    fields={
      Array [
        Object {
          "name": "name",
          "placeholder": "Name..",
          "title": "User Name",
        },
        Object {
          "name": "hiatuses",
          "options": Object {
            "component": [Function],
            "type": "custom",
          },
          "title": "Hiatuses",
        },
      ]
    }
    handleClose={[Function]}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "_type": "user",
        "friends": Array [
          "uid0",
          "uid1",
          "uid2",
          "uid3",
          "uid4",
        ],
        "hiatuses": Array [],
        "name": "Mr. User",
        "preferences": Object {
          "activityCalendar": Object {
            "friends": Array [],
            "includeWorkouts": true,
            "mode": "User only",
            "splitWorkouts": true,
          },
        },
        "uid": "myUid",
      }
    }
    show={false}
    title="Edit User"
    validateState={[Function]}
  >
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      dialogAs={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "ModalDialog",
          "render": [Function],
        }
      }
      enforceFocus={true}
      keyboard={true}
      onHide={[Function]}
      restoreFocus={true}
      show={false}
    >
      <Modal
        autoFocus={true}
        backdrop={true}
        backdropTransition={[Function]}
        containerClassName="modal-open"
        enforceFocus={true}
        keyboard={true}
        manager={
          BootstrapModalManager {
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "modals": Array [],
            "scrollbarSize": 0,
          }
        }
        onEnter={[Function]}
        onEntering={[Function]}
        onEscapeKeyDown={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onHide={[Function]}
        renderBackdrop={[Function]}
        renderDialog={[Function]}
        restoreFocus={true}
        show={false}
        transition={[Function]}
      />
    </Modal>
    <ConfirmCancelModal
      handleConfirm={[Function]}
      hideModal={[Function]}
      modalTitle="Delete session?"
      showModal={false}
    >
      <Modal
        animation={true}
        autoFocus={true}
        backdrop={true}
        dialogAs={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "ModalDialog",
            "render": [Function],
          }
        }
        enforceFocus={true}
        keyboard={true}
        onHide={[Function]}
        restoreFocus={true}
        show={false}
      >
        <Modal
          autoFocus={true}
          backdrop={true}
          backdropTransition={[Function]}
          containerClassName="modal-open"
          enforceFocus={true}
          keyboard={true}
          manager={
            BootstrapModalManager {
              "containers": Array [],
              "data": Array [],
              "handleContainerOverflow": true,
              "hideSiblingNodes": true,
              "modals": Array [],
              "scrollbarSize": 0,
            }
          }
          onEnter={[Function]}
          onEntering={[Function]}
          onEscapeKeyDown={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onHide={[Function]}
          renderBackdrop={[Function]}
          renderDialog={[Function]}
          restoreFocus={true}
          show={false}
          transition={[Function]}
        />
      </Modal>
    </ConfirmCancelModal>
  </EntityModal>
  <EntityModal
    fields={
      Array [
        Object {
          "name": "userValue",
          "placeholder": "User...",
          "title": "User ID or Name",
        },
      ]
    }
    handleClose={[Function]}
    handleSubmit={[Function]}
    show={false}
    title="Add friend"
    validateState={[Function]}
  >
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      dialogAs={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "ModalDialog",
          "render": [Function],
        }
      }
      enforceFocus={true}
      keyboard={true}
      onHide={[Function]}
      restoreFocus={true}
      show={false}
    >
      <Modal
        autoFocus={true}
        backdrop={true}
        backdropTransition={[Function]}
        containerClassName="modal-open"
        enforceFocus={true}
        keyboard={true}
        manager={
          BootstrapModalManager {
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "modals": Array [],
            "scrollbarSize": 0,
          }
        }
        onEnter={[Function]}
        onEntering={[Function]}
        onEscapeKeyDown={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onHide={[Function]}
        renderBackdrop={[Function]}
        renderDialog={[Function]}
        restoreFocus={true}
        show={false}
        transition={[Function]}
      />
    </Modal>
    <ConfirmCancelModal
      handleConfirm={[Function]}
      hideModal={[Function]}
      modalTitle="Delete session?"
      showModal={false}
    >
      <Modal
        animation={true}
        autoFocus={true}
        backdrop={true}
        dialogAs={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "ModalDialog",
            "render": [Function],
          }
        }
        enforceFocus={true}
        keyboard={true}
        onHide={[Function]}
        restoreFocus={true}
        show={false}
      >
        <Modal
          autoFocus={true}
          backdrop={true}
          backdropTransition={[Function]}
          containerClassName="modal-open"
          enforceFocus={true}
          keyboard={true}
          manager={
            BootstrapModalManager {
              "containers": Array [],
              "data": Array [],
              "handleContainerOverflow": true,
              "hideSiblingNodes": true,
              "modals": Array [],
              "scrollbarSize": 0,
            }
          }
          onEnter={[Function]}
          onEntering={[Function]}
          onEscapeKeyDown={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onHide={[Function]}
          renderBackdrop={[Function]}
          renderDialog={[Function]}
          restoreFocus={true}
          show={false}
          transition={[Function]}
        />
      </Modal>
    </ConfirmCancelModal>
  </EntityModal>
  <ActivityCalendarSettingsModal
    friends={
      Array [
        Object {
          "key": "uid0-key",
          "value": Object {
            "_type": "user",
            "friends": Array [],
            "hiatuses": Array [],
            "name": "user 0",
            "preferences": Object {
              "activityCalendar": Object {
                "friends": Array [],
                "includeWorkouts": true,
                "mode": "User only",
                "splitWorkouts": true,
              },
            },
            "uid": "uid0",
          },
        },
        Object {
          "key": "uid1-key",
          "value": Object {
            "_type": "user",
            "friends": Array [],
            "hiatuses": Array [],
            "name": "user 1",
            "preferences": Object {
              "activityCalendar": Object {
                "friends": Array [],
                "includeWorkouts": true,
                "mode": "User only",
                "splitWorkouts": true,
              },
            },
            "uid": "uid1",
          },
        },
        Object {
          "key": "uid2-key",
          "value": Object {
            "_type": "user",
            "friends": Array [],
            "hiatuses": Array [],
            "name": "user 2",
            "preferences": Object {
              "activityCalendar": Object {
                "friends": Array [],
                "includeWorkouts": true,
                "mode": "User only",
                "splitWorkouts": true,
              },
            },
            "uid": "uid2",
          },
        },
        Object {
          "key": "uid3-key",
          "value": Object {
            "_type": "user",
            "friends": Array [],
            "hiatuses": Array [],
            "name": "user 3",
            "preferences": Object {
              "activityCalendar": Object {
                "friends": Array [],
                "includeWorkouts": true,
                "mode": "User only",
                "splitWorkouts": true,
              },
            },
            "uid": "uid3",
          },
        },
        Object {
          "key": "uid4-key",
          "value": Object {
            "_type": "user",
            "friends": Array [],
            "hiatuses": Array [],
            "name": "user 4",
            "preferences": Object {
              "activityCalendar": Object {
                "friends": Array [],
                "includeWorkouts": true,
                "mode": "User only",
                "splitWorkouts": true,
              },
            },
            "uid": "uid4",
          },
        },
      ]
    }
    handleClose={[Function]}
    handleSubmit={[Function]}
    show={false}
    user={
      Object {
        "_type": "user",
        "friends": Array [
          "uid0",
          "uid1",
          "uid2",
          "uid3",
          "uid4",
        ],
        "hiatuses": Array [],
        "name": "Mr. User",
        "preferences": Object {
          "activityCalendar": Object {
            "friends": Array [],
            "includeWorkouts": true,
            "mode": "User only",
            "splitWorkouts": true,
          },
        },
        "uid": "myUid",
      }
    }
  >
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      dialogAs={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "ModalDialog",
          "render": [Function],
        }
      }
      enforceFocus={true}
      keyboard={true}
      onHide={[Function]}
      restoreFocus={true}
      show={false}
    >
      <Modal
        autoFocus={true}
        backdrop={true}
        backdropTransition={[Function]}
        containerClassName="modal-open"
        enforceFocus={true}
        keyboard={true}
        manager={
          BootstrapModalManager {
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "modals": Array [],
            "scrollbarSize": 0,
          }
        }
        onEnter={[Function]}
        onEntering={[Function]}
        onEscapeKeyDown={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onHide={[Function]}
        renderBackdrop={[Function]}
        renderDialog={[Function]}
        restoreFocus={true}
        show={false}
        transition={[Function]}
      />
    </Modal>
  </ActivityCalendarSettingsModal>
</UserPage>
`;
